### PR TITLE
Di and compose

### PR DESCRIPTION
**Pull request by pasha-dektev**
1. Integrated Dagger
2. Implemented UI using Jetpack Compose

### CODE REVIEW

File: app/src/main/java/com/lightricks/feedexercise/FeedApp.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.lightricks.feedexercise
+
+import android.app.Application
+import android.content.Context
+import com.lightricks.feedexercise.di.AppComponent
+import com.lightricks.feedexercise.di.DaggerAppComponent
+
+internal class FeedApp : Application() {
+    lateinit var appComponent: AppComponent
+        private set
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/FeedApp.kt:11]: You must synchronize the visibility with the other threads.

File: app/src/main/java/com/lightricks/feedexercise/FeedApp.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            .context(this)
+            .build()
+    }
+}
+
+internal fun Context.appComponent(): AppComponent {
+    return when (this) {
+        is FeedApp -> appComponent
+        else -> this.applicationContext.appComponent()
+    }
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/FeedApp.kt]: Since you are assuming that the applicationContext is a FeedApp then make and explicit cast and remove the recursion. So that if the cast failed we crash. Otherwise we will get in such a case in an infinite recursion. 

```kotlin
internal fun Context.appComponent(): AppComponent {
    return (applicationContext as FeedApp).appComponent
}
```

File: app/src/main/java/com/lightricks/feedexercise/data/remote/RemoteFeedDataSource.kt [hunk 3]
```diff
@@ -9,13 +9,16 @@ import com.lightricks.feedexercise.util.result.toFailureResult
...
+import javax.inject.Named
 
 interface RemoteFeedDataSource {
     suspend fun getFeed(): Result<FeedError, FeedSuccess>
 }
 
-internal class RetrofitFeedDataSource(
+internal class RetrofitFeedDataSource @Inject constructor(
     private val apiService: FeedApiService,
+    @Named("ThumbnailImageBaseUrl")
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/remote/RemoteFeedDataSource.kt:23]: minor: In general in a real application we will prefer to work with a custom qualifier so that will avoid us to deal with strings.

File: app/src/main/java/com/lightricks/feedexercise/di/Di.kt [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    }
+}
+
+@Module
+internal interface FeedBindModule {
+    @Binds
+    @Singleton
+    fun bindRetrofitFeedDataSource(
+        retrofitFeedDataSource: RetrofitFeedDataSource
+    ): RemoteFeedDataSource
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/di/Di.kt]: We can put the scope (Singleton in this case) on the implementation since in general it depends on the implementation details.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedAdapter.kt [hunk 5]
```diff
@@ -1,73 +0,0 @@
-package com.lightricks.feedexercise.ui.feed
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedAdapter.kt:1]: This change is not related to the DI commit.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt [hunk 6]
```diff
@@ -1,71 +1,77 @@
...
-        dataBinding = DataBindingUtil.inflate(inflater, R.layout.feed_fragment, container, false)
-        setupViewModel()
-        setupViews()
-        return dataBinding.root
-    }
-
-    private fun setupViewModel() {
-        viewModel = ViewModelProvider(this, FeedViewModelFactory())
-            .get(FeedViewModel::class.java)
+class FeedFragment : Fragment(R.layout.feed_fragment) {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt]: Same for this file, there're a lot of changes that are not related to the DI commit.

File: app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt [hunk 7]
```diff
@@ -1,15 +1,77 @@
...
-        setContentView(R.layout.main_activity)
+        setContent {
+            MaterialTheme {
+                FeedScreen(viewModel)
+            }
+        }
+    }
+
+    @Composable
+    fun FeedScreen(viewModel: FeedViewModel) {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: private?

File: app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt [hunk 8]
```diff
@@ -1,15 +1,77 @@
...
+                        modifier = Modifier
+                            .align(Alignment.CenterHorizontally)
+                            .padding(top = 4.dp),
+                        onClick = { viewModel.onRefresh() }
+                    ) {
+                        Text(text = "Reload")
+                    }
+                }
+            }
+        }
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: remove?

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/Error.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.stringResource
+import com.lightricks.feedexercise.R
+import com.lightricks.feedexercise.ui.feed.model.ViewState
+
+@Composable
+internal fun ErrorState(
+    state: ViewState.Error,
+    viewModel: FeedViewModel
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/Error.kt]: I have never worked with compose but I don't think it is a good idea to couple the component with the view model.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 10]
```diff
@@ -1,58 +1,69 @@
...
+    )
 
-    fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
+    val viewState: StateFlow<ViewState> = engine.output
+        .map { state -> createViewState(state) }
+        .stateIn(
+            scope = viewModelScope,
+            started = SharingStarted.WhileSubscribed(5000),
+            initialValue = ViewState.Loading
+        )
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: I am not sure I understand why do we need this operator.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 11]
```diff
@@ -1,58 +1,69 @@
...
-    }
-
-    private fun updateState(transform: State.() -> State) {
-        stateInternal.value = transform(getState())
+    /**
+     * Converts domain state to view state
+     * @param state domain state
+     * @return view state
+     */
+    private fun createViewState(state: State): ViewState {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: It is rather a mapping method.

**Review by nissimLightricks — CHANGES_REQUESTED** [review]
Excellent job!!! In general a commit should contains only one thing/context of change. Here the commit contains two different things. DI + Compose. 

Round.

File: app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt (hunk 8 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: Why? Its only a compose preview - it only shows you the result without launching the app - it does not affect anything

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/Error.kt (hunk 9 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/Error.kt]: You are right - it should be callback

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt (hunk 10 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: Mainly it is for unsubscribing when there are no observers - you can read a great post about it (https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda)

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt (hunk 11 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: Yes. It is. I'll rename it

File: app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt (hunk 8 above)
Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: This is good during debug not to merge in the code base.

Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: Its like using tools:xxx inside layout. But doesnt matter - I will remove it

Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/ui/MainActivity.kt]: IMO this pollute the code, I am not sure how we should work with compose, but for sure I will not want to see a code that is there for debug as part of the code base. 

And no I don't think it is the same, in xml it is very clear what is for debug and what is not, here you add code that looks exactly like a regular compose function that can be use by mistake in production.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt (hunk 10 above)
Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: But here it is a StateFlow (hot flow) so I don't understand the point.. You are stopping what exactly?

File: app/src/main/java/com/lightricks/feedexercise/data/remote/RemoteFeedDataSource.kt (hunk 3 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/data/remote/RemoteFeedDataSource.kt:23]: Noted

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt (hunk 10 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: Yeah - I know. WhileSubscribed will stop the underlying producer when there are no active observers.
The compose function observes stateFlow. When we call another compose function there is no observer on state flow, so it will stop emitting states after 5000 ms

File: app/src/main/java/com/lightricks/feedexercise/di/Di.kt (hunk 4 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/di/Di.kt]: I dont understand what you mean... Put scope where? On RetrofitFeedDataSource?

File: app/src/main/java/com/lightricks/feedexercise/FeedApp.kt (hunk 1 above)
Comment (pasha-dektev) [reply app/src/main/java/com/lightricks/feedexercise/FeedApp.kt:11]: I dont get it. It allows only reading - why do we have to synchronize?

File: app/src/main/java/com/lightricks/feedexercise/di/Di.kt (hunk 4 above)
Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/di/Di.kt]: Yes, exactly on RetrofitFeedDataSource

File: app/src/main/java/com/lightricks/feedexercise/FeedApp.kt (hunk 1 above)
Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/FeedApp.kt:11]: Read about java memory model it is an important subject. https://jenkov.com/tutorials/java-concurrency/java-memory-model.html

File: app/src/main/java/com/lightricks/feedexercise/FeedApp.kt [hunk 12]
```diff
@@ -20,6 +21,6 @@ internal class FeedApp : Application() {
 internal fun Context.appComponent(): AppComponent {
     return when (this) {
         is FeedApp -> appComponent
-        else -> this.applicationContext.appComponent()
+        else -> (applicationContext as FeedApp).appComponent
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/FeedApp.kt]: The when is not needed here. You can replace all the content of the method by `return (applicationContext as FeedApp).appComponent`

File: app/src/main/java/com/lightricks/feedexercise/network/UrlProvider.kt [hunk 13]
```diff
@@ -9,13 +10,15 @@ interface UrlProvider {
 /**
  * Base url provider for Retrofit
  */
+@Singleton
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/UrlProvider.kt]: Since this implementation doesn't hold any state then there's no reason of having it as singleton. It is even "wasteful" since it use only a small scope/context. So we should not prevent the GC to collect it when no more in use.

File: app/src/main/java/com/lightricks/feedexercise/network/UrlProvider.kt [hunk 14]
```diff
@@ -9,13 +10,15 @@ interface UrlProvider {
...
 class RetrofitBaseUrlProvider @Inject constructor() : UrlProvider {
     override val url: String = "https://assets.swishvideoapp.com/"
 }
 
 /**
  * Base url provider for Image Loading
  */
+@Singleton
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/UrlProvider.kt]: Same for this one.

**Review by nissimLightricks — CHANGES_REQUESTED** [review]
Good job. F&S.

**Review by nissimLightricks — APPROVED** [review]
